src/src/PluginStructs/P122_data_struct.cpp

```diff
@@ -0,0 +1,568 @@
+
+//////////////////////////////////////////////////////////////////////////////////////////////////
+uint8_t P122_data_struct::crc8(const uint8_t *data, uint8_t len)
+{
+  // CRC-8 formula from page 14 of SHT spec pdf
+  // Sensirion_Humidity_Sensors_SHT2x_CRC_Calculation.pdf
+  const uint8_t POLY = 0x31;
+  uint8_t crc = 0x00;
+
+  for (uint8_t j = len; j; --j)
+  {
+    crc ^= *data++;
+
+    for (uint8_t i = 8; i; --i)
+    {
+      crc = (crc & 0x80) ? (crc << 1) ^ POLY : (crc << 1);
+    }
+  }
+  return crc;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////
+bool P122_data_struct::writeCmd(uint8_t cmd)
+{
+  return I2C_write8(_i2caddr, cmd);
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////
+bool P122_data_struct::writeCmd(uint8_t cmd, uint8_t value)
+{
+  return I2C_write8_reg(_i2caddr, cmd, value);
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////
+bool P122_data_struct::readBytes(uint8_t n, uint8_t *val, uint8_t maxDuration)
+{
+  // TODO check if part can be delegated to the I2C_access libraray from ESPeasy
+  Wire.requestFrom(_i2caddr, (uint8_t) n);
+  uint32_t start = millis();
+  while (Wire.available() < n)
+  {
+    if (millis() - start > maxDuration)
+    {
+      return false;
+    }
+    yield();
+  }
+
+  for (uint8_t i = 0; i < n; i++)
+  {
+    val[i] = Wire.read();
+  }
+  return true;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////
+// Send Temperature measurement command to device
+bool P122_data_struct::requestTemperature()
+{
+  return writeCmd(SHT2x_GET_TEMPERATURE_NO_HOLD);
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////
+// Sent Humidity measurement command to device
+bool P122_data_struct::requestHumidity()
+{
+  return writeCmd(SHT2x_GET_HUMIDITY_NO_HOLD);
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////
+// Read temperature measurement results from device
+bool P122_data_struct::readTemperature()
+{
+  uint8_t buffer[3];
+
+  if (!readBytes(3, (uint8_t*) &buffer[0], 90))
+  {
+    return false;
+  }
+  if (crc8(buffer, 2) == buffer[2])
+  {
+    _rawTemperature  = buffer[0] << 8;
+    _rawTemperature += buffer[1];
+    _rawTemperature &= 0xFFFC;
+  }
+  return true;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////
+// Read humidity measurement results from device
+bool P122_data_struct::readHumidity()
+{
+  uint8_t buffer[3];
+
+  if (!readBytes(3, (uint8_t*) &buffer[0], 30))
+  {
+    return false;
+  }
+  if (crc8(buffer, 2) == buffer[2])
+  {
+    _rawHumidity  = buffer[0] << 8;
+    _rawHumidity += buffer[1];
+    _rawHumidity &= 0xFFFC;
+  }
+   return true;
+}
```

Comment:
This code is a duplicate of `readTemperature()`, except using register `30` instead of `90`. Could probably be a single function with that as an argument?

Reply:
Didn't see this myself... Started with borrowed code and restructured it until almost the same code.